app/controllers/products_controller.rb

```diff
@@ -1,29 +1,76 @@
 class ProductsController < ApplicationController
-  before_action :set_product, only: [:show, :add_to_cart]
+  before_action :set_product, only: [:show, :edit, :update, :retire]
+  before_action :set_seller, only: [:new, :create, :seller]
+  before_action :require_seller_login, only: [:new, :update, :edit, :create, :add_categories, :retire]
+  before_action :require_seller_login_if_retired, only: [:show]
 
-  def add_to_cart
-    session[:order_id] ||= Order.create.id
+  def index
+    @products = Product.active
+  end
+
+  def show
+    @reviews = @product.reviews
+    @average_rating = @product.average_rating
+    @product_categories = @product.categories
+  end
 
-    # this should perhaps be a method in the order model
-    unless Order.find(session[:order_id]).already_has_product?(@product.id)
-      OrderItem.create(product_id: @product.id, order_id: session[:order_id], quantity_ordered: 1)
+  def new
+    @product = Product.new
+  end
+
+  def edit
+    @categories = Category.all
+  end
+
+  def seller; end
+
+  def create
+    @product = Product.new(create_params)
+    @product.seller_id = params[:seller_id]
+    if @product.save
+      redirect_to product_path(@product)
+    else
+      flash.now[:errors] = @product.errors
+      render :new
     end
+  end
 
-    redirect_to cart_path # this should redirect to product show page
+  def update
+    if @product.update(create_params)
+      redirect_to seller_products_path(@product.seller_id)
+    else
+      flash.now[:errors] = @product.errors
+      render :edit
+    end
   end
 
-  def index
-    @products = Product.all
+  def retire
+    @product.retire!
+    redirect_to product_path(@product)
   end
 
-  def show
-    @reviews = Product.find(params[:id]).reviews
-    @average_rating = Product.average_rating(params[:id])
+  def add_categories
+    @product = Product.find(params[:product_id])
+    params[:category_id].each do |category_id|
+      category = Category.find(category_id)
+      @product.categories << category
+    end
+    redirect_to seller_products_path(@product.seller_id)
   end
 
+
   private
+    def set_product
+      @product = Product.find(params[:id])
+    end
 
-  def set_product
-    @product = Product.find(params[:id])
-  end
+    def create_params
+      params.require(:product).permit(:name, :price, :stock, :description, :photo_url)
+    end
+
+    def require_seller_login_if_retired
```

Comment:
This method seems unnecessary since it's only used in the show and it could be re-written in a single-line `if `statement